locator/token_metadata.cc

```diff
@@ -1175,30 +1199,47 @@ void generic_token_metadata<NodeId>::del_replacing_endpoint(NodeId existing_node
 
 template <typename NodeId>
 future<generic_token_metadata<NodeId>> generic_token_metadata<NodeId>::clone_async() const noexcept {
-    return _impl->clone_async().then([] (std::unique_ptr<token_metadata_impl<NodeId>> impl) {
-        return make_ready_future<generic_token_metadata>(std::move(impl));
-    });
+    if constexpr (std::is_same_v<NodeId, gms::inet_address>) {
+        co_return !holds_alternative<std::monostate>(_new_value)
+                  ? generic_token_metadata(co_await _impl->clone_async(), co_await get_new()->clone_async())
+                  : generic_token_metadata(co_await _impl->clone_async());
+    } else {
+        co_return generic_token_metadata(co_await _impl->clone_async());
+    }
 }
 
 template <typename NodeId>
 future<generic_token_metadata<NodeId>>
 generic_token_metadata<NodeId>::clone_only_token_map() const noexcept {
-    return _impl->clone_only_token_map().then([] (std::unique_ptr<token_metadata_impl<NodeId>> impl) {
-        return generic_token_metadata(std::move(impl));
-    });
+    if constexpr (std::is_same_v<NodeId, gms::inet_address>) {
+        co_return !holds_alternative<std::monostate>(_new_value)
+                  ? generic_token_metadata(co_await _impl->clone_only_token_map(), co_await get_new()->clone_only_token_map())
+                  : generic_token_metadata(co_await _impl->clone_only_token_map());
+    } else {
+        co_return generic_token_metadata(co_await _impl->clone_only_token_map());
+    }
 }
 
 template <typename NodeId>
 future<generic_token_metadata<NodeId>>
 generic_token_metadata<NodeId>::clone_after_all_left() const noexcept {
-    return _impl->clone_after_all_left().then([] (std::unique_ptr<token_metadata_impl<NodeId>> impl) {
-        return generic_token_metadata(std::move(impl));
-    });
+    if constexpr (std::is_same_v<NodeId, gms::inet_address>) {
+        co_return !holds_alternative<std::monostate>(_new_value)
+                  ? generic_token_metadata(co_await _impl->clone_after_all_left(), co_await get_new()->clone_after_all_left())
+                  : generic_token_metadata(co_await _impl->clone_after_all_left());
+    } else {
+        co_return generic_token_metadata(co_await _impl->clone_after_all_left());
+    }
 }
 
 template <typename NodeId>
 future<> generic_token_metadata<NodeId>::clear_gently() noexcept {
-    return _impl->clear_gently();
+    co_await _impl->clear_gently();
+    if constexpr (std::is_same_v<NodeId, gms::inet_address>) {
+        if (holds_alternative<lw_shared_ptr<token_metadata2>>(_new_value)) {
```

Comment:
Is it even correct to call this function of the immutable variant is stored? With the current implementation, you would not clear anything, which doesn't sound right.